src/services/saml-config-service.js

```diff
@@ -44,28 +39,81 @@ export default class SamlConfigService {
     }
   }
 
-  resolveMetadata(xmlDoc, entityId) {
-    const descriptorRootXpath = `sm:EntityDescriptor[@entityID="${entityId}"]/sm:IDPSSODescriptor`;
+  resolveMetadata(xml, entityId) {
+    return new Promise((resolve, reject) => {
+      let entryPoint = null;
+      let certificates = [];
+      let currentEntityIdValue = null;
+      let readNextTextAsCertificate = false;
 
-    const certificatesXpath = `//${descriptorRootXpath}/sm:KeyDescriptor//ds:X509Certificate`;
-    const certificates = xpathSelect(certificatesXpath, xmlDoc).map(c => c.textContent.trim() || null).filter(c => c);
+      const parser = sax.parser(true, { xmlns: true });
 
-    const entryPointXpath = `//${descriptorRootXpath}/sm:SingleSignOnService[@Binding="${REDIRECT_BINDING}"]/@Location`;
-    const entryPoint = xpathSelect(entryPointXpath, xmlDoc)[0]?.value.trim() || null;
+      parser.onopentag = node => {
+        if (node.local === 'X509Certificate' && node.uri === XMLSN_URI_DS) {
+          readNextTextAsCertificate = currentEntityIdValue === entityId;
+          return;
+        }
+
+        readNextTextAsCertificate = false;
+
+        if (node.local === 'SingleSignOnService' && node.uri === XMLNS_URI_SM) {
+          if (currentEntityIdValue !== entityId) {
+            return;
+          }
+
+          const { binding, location } = Object.values(node.attributes).reduce((data, attr) => {
```

Comment:
If there aren't alot of attributes to parse, I think the code would be clearer with

```
const redirectBinding = Object.values(node.attributes).find(attr => attr.local === 'Binding' && attr.uri === '' && attr.value === REDIRECT_BINDING);
const location = Object.values(node.attributes).find(attr => attr.local === 'Location' && attr.uri === '' && !!attr.value);
```
otherwise maybe
```
let foundRedirectBinding;
let foundLocation;
Object.values(node.attributes).forEach(attr => {
   if (attr.local === 'Binding' && attr.uri === '' && attr.value === REDIRECT_BINDING) {
       foundRedirectBinding = true;
   }
   if (attr.local === 'Location' && attr.uri === '' && !!attr.value) {
       foundLocation = true;
   }
})
```

Passing `data` around in the `reduce` I find makes it a bit harder to follow.
What do you think?

Reply:
Much clearer now also with the simplified early return! 👏